### ISSUE TITLE
use first response from Esri service to validate in hooked, instead of all features.

### ISSUE THREAD

**Issue by slibby** [openaddresses/machine#194]
From my reading on how the hooked CI uses machine, when downloading shapefiles and running them through hooked, it samples 6 features from them and uses that to validate the dataset. This is pretty speedy and :+1:

When querying an Esri service, hooked queries ALL features from the service before sampling from it, checking them and then choking on something simple like a field name error, could this be changed to only download 1,000 features (the default query amount) or fewer when it's in "hooked mode"? This would speed up the CI responses for hooked and also reduce load/data xfer on the ci machines substantially I think.

I was going to take a crack at it in a PR but don't know enough about how hooked invokes cache.py vs. how the weekly load invokes it, so I didn't want to add something that might break it.

Thoughts?

**Comment by migurski** [comment]
I believe the way this happens is that ESRI sources are first converted to GeoJSON, then that GeoJSON is sampled. [Here’s the sampling](https://github.com/openaddresses/machine/blob/2.1.2/openaddr/conform.py#L135) showing the limited set of input formats, and [this looks to be the ESRI download](https://github.com/openaddresses/machine/blob/2.1.2/openaddr/cache.py#L273) showing how the whole thing is walked and converted to raw geometries then GeoJSON.

I believe one of the issues we’ve had with ESRI sources is that it’s only recent versions of Feature Server which allow feature enumeration, so @iandees and @ingalls worked on a recursive descent process.

**Comment by slibby** [comment]
Is there a way that cache.py "knows" that it's running in hooked/ci mode rather than batch mode? If that was an argument somewhere, you could break out of the esri loop after 1000 features are successfully downloaded, and then continue with the rest.

**Comment by migurski** [comment]
Not really. The `hooked/ci` mode isn’t intended to merely check whether the source is okay, it actually performs the full run and stores the full results. A page [like this](http://ci.openaddresses.io/jobs/0bb80c50-235e-4347-9b91-f2bb7826837f) has complete downloadable results available.

**Comment by NelsonMinar** [comment]
I think it would be a good improvement for the caching code to always only download a minimal amount of data, right? Doesn't matter if CI or whatever. The big drawback I see is that entails a whole second download process.

**Comment by migurski** [comment]
I’m not sure if I understand—I think (would need to verify) that the caching code is the only download in the chain, and the conform code operates on a locally-cached copy.

**Comment by NelsonMinar** [comment]
Sorry, what I had in mind is a special downloader for the sampler that only downloads a few records. Not sure if it's really worth a special effort though.

**Comment by migurski** [comment]
Ah, hm, yeah. Hm.

Hm.
